lib/de.ml

```diff
@@ -2605,6 +2887,1050 @@ module Def = struct
+
+    let num_literal_observation_types = 8
+    let num_match_observation_types = 2
+
+    let num_observation_types =
+      num_literal_observation_types + num_match_observation_types
+
+    type block_split_stats = {
+        new_observations: int array
+      ; observations: int array
+      ; mutable num_new_observations: int
+      ; mutable num_observations: int
+    }
+
+    let split_stats =
+      {
+        new_observations= Array.make num_observation_types 0
+      ; observations= Array.make num_observation_types 0
+      ; num_new_observations= 0
+      ; num_observations= 0
+      }
+
+    type lens = {mutable best: int; mutable nice: int; mutable max: int}
+
+    let blocktype_uncompressed = 0
+    let blocktype_static_huffman = 1
+    let blocktype_dynamic_huffman = 2
+
+    let init_output i o =
+      {
+        i
+      ; i_pos= 0
+      ; i_base= 0
+      ; i_len= bigstring_length i
+      ; o
+      ; o_pos= 0
+      ; o_len= bigstring_length o - end_padding
+      ; hold= 0
+      ; bits= 0
+      }
+
+    let length_slot_base =
+      [|
+         3; 4; 5; 6; 7; 8; 9; 10; 11; 13; 15; 17; 19; 23; 27; 31; 35; 43; 51; 59
+       ; 67; 83; 99; 115; 131; 163; 195; 227; 258
+      |]
+
+    let extra_length_bits =
+      [|
+         0; 0; 0; 0; 0; 0; 0; 0; 1; 1; 1; 1; 2; 2; 2; 2; 3; 3; 3; 3; 4; 4; 4; 4
+       ; 5; 5; 5; 5; 0
+      |]
+
+    let offset_slot_base =
+      [|
+         1; 2; 3; 4; 5; 7; 9; 13; 17; 25; 33; 49; 65; 97; 129; 193; 257; 385; 513
+       ; 769; 1025; 1537; 2049; 3073; 4097; 6145; 8193; 12289; 16385; 24577
+      |]
+
+    let extra_offset_bits =
+      [|
+         0; 0; 0; 0; 1; 1; 2; 2; 3; 3; 4; 4; 5; 5; 6; 6; 7; 7; 8; 8; 9; 9; 10; 10
+       ; 11; 11; 12; 12; 13; 13
+      |]
+
+    let length_slot =
+      [|
+         0; 0; 0; 0; 1; 2; 3; 4; 5; 6; 7; 8; 8; 9; 9; 10; 10; 11; 11; 12; 12; 12
+       ; 12; 13; 13; 13; 13; 14; 14; 14; 14; 15; 15; 15; 15; 16; 16; 16; 16; 16
+       ; 16; 16; 16; 17; 17; 17; 17; 17; 17; 17; 17; 18; 18; 18; 18; 18; 18; 18
+       ; 18; 19; 19; 19; 19; 19; 19; 19; 19; 20; 20; 20; 20; 20; 20; 20; 20; 20
+       ; 20; 20; 20; 20; 20; 20; 20; 21; 21; 21; 21; 21; 21; 21; 21; 21; 21; 21
+       ; 21; 21; 21; 21; 21; 22; 22; 22; 22; 22; 22; 22; 22; 22; 22; 22; 22; 22
+       ; 22; 22; 22; 23; 23; 23; 23; 23; 23; 23; 23; 23; 23; 23; 23; 23; 23; 23
+       ; 23; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24
+       ; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 24; 25; 25; 25
+       ; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25
+       ; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 25; 26; 26; 26; 26; 26; 26; 26
+       ; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26; 26
+       ; 26; 26; 26; 26; 26; 26; 26; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27
+       ; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27; 27
+       ; 27; 27; 28
+      |]
```

Comment:
Can be replaced by `_length` minor extra prepended `0`. Don't be afraid to update `_lenght` and the stream API to use the same array.